Runtime/Managers/Implementations/RumbleManager.cs

```diff
@@ -0,0 +1,65 @@
+﻿using GameplayIngredients;
+using UnityEngine;
+using UnityEngine.InputSystem;
```

Comment:
FYI, Gameplay Ingredients both supports old and new input system, so you should encapsulate every feature that is compatible with only one system by `#if` statements. There are some examples in 

https://github.com/peeweek/net.peeweek.gameplay-ingredients/blob/master/Runtime/LevelScripting/Events/OnInputActionEvent.cs

and 

https://github.com/peeweek/net.peeweek.gameplay-ingredients/blob/master/Runtime/LevelScripting/Events/OnKeyDownEvent.cs

(Basically `#if ENABLE_INPUT_SYSTEM` if the new input system package is enabled and `#if ENABLE_LEGACY_INPUT_MANAGER` if the old input system is enabled

Reply:
Hi Thomas! Sorry! I just read now your messages :D 

I'm coming from Unreal Engine so I need a little bit time to understand how Unity works, actually I'm interested in low level render pipeline features. I will break the PR so you can review. 

[quote]
Inclusion List : is it mutually exclusive with the exclusion list? (In that case, wouldn't it be better to use a single list and a behavior selector (include from list/exclude from list)? I'm thinking about manually managing these two lists, as a manager could be included then excluded, which would make the behavior cumbersome )

Yes, actually I just use the inclusion list (I didn't remove the exclusion list because you are the boss and I would like to talk with you about it). The inclusion list in my opinion gives you a big advantage: loading order. I wrote for example a tooltip manager that needs to get a component from the UIEventManager during initialization. With the inclusion list you can choose the sequence of loading and access the managers already loaded with Manager.TryGet<>().  

[quote]
Per-scene managers : I really can´t understand the purpose of your addition: As the AutoCreateAll() method happens only once before the first scene loads. It will never be called when you load other scenes. Remember: Managers are singletons which object lifecycle transcends scene loads/unloads. Game objects bound to a scene should be embedded in this scene and not treated as singleton Managers.

The per-scene managers/prefabs are useful during testing/development purpose. If I want to tweak the VirtualCameraManager prefab in a test-scene without changing the "main" prefab in Assets/Resources I can do it. I found it very useful during development.  

[quote]
Also, you seem to have included a RenderPipelineSwitcher option whch only switches between URP and Built-In : this one is out-of scope for the pull request, please remove it.

Sure, I will remove it. I read on your CHANGELOG that the custom toolbar works also with 2021.2. I downloaded the 2021.2.6f1 but I can't see the toolbar anymore, should I activate it manually ? 

[quote]
Please see my comments about splitting your PR into multiple PRs (one for each improvement), it's starting to grow and will become harder to merge.

Yes, I will... starting from now, promise :) 

Have a nice day and thank you for your comments.
Giuseppe

Reply:
>The inclusion list in my opinion gives you a big advantage: loading order. I wrote for example a tooltip manager that needs to get a component from the UIEventManager during initialization. With the inclusion list you can choose the sequence of loading and access the managers already loaded with Manager.TryGet<>().

Interesting, you had the same manager dependency issue as I had a while ago. My current plans for loading dependencies was to add an optional `ManagerDependsOnAttribute` in the manager class definition, to define all the required dependencies for a manager, and thus perform reordering the list before loading the prefabs. This would probably help solve exclusion list issues, such as missing managers. But the purpose of automatic dependency would be to release that constraint from the users (as they are probably not aware of these dependencies in the first place)

for example: `[ManagerDependsOn(typeof(UIEventManager),typeof(GameManager))]`

> The per-scene managers/prefabs are useful during testing/development purpose. If I want to tweak the VirtualCameraManager prefab in a test-scene without changing the "main" prefab in Assets/Resources I can do it. I found it very useful during development.

Oh, so it seems that it's an editor-only feature? If I understand correctly your use case, you want to do local changes before applying them to the main prefab? In that case I have some suggestions/concerns:
- When we are in a "scene overridden" mode, it can be difficult to tell which managers are not using the default prefabs, thus behaviour could be different between in-editor play mode and built players. I am a bit concerned about that as it would probably require a clear feedback to view these changes, and also be able to test it in both conditions.
- Performing editor-only changes / testing requires if possible as few changes to the main runtime code. While it's not always possible (for example, I had to add a `EditorLogic` to handle correctly Play from Here). The rule of thumb is to leave the runtime code unaffected, for example by adding extra editor code between `#if UNITY_EDITOR` directives.

> Sure, I will remove it. I read on your CHANGELOG that the custom toolbar works also with 2021.2. I downloaded the 2021.2.6f1 but I can't see the toolbar anymore, should I activate it manually ?

The new overlay system has probably toggled off the visibility of the additional toolbar by default, you can display the visibility options by pressing spacebar and selecting the _Gameplay Ingredients (Custom)_ toolbar from the menu (I made it a separate one now)

Please note that your additions are pretty interesting, I am absolutely not against these but these require a bit more

Reply:
> Interesting, you had [the same manager dependency issue as I had a while ag](https://github.com/peeweek/net.peeweek.gameplay-ingredients/issues/57)o. My current plans for loading dependencies was to add an optional `ManagerDependsOnAttribute` in the manager class definition, to define all the required dependencies for a manager, and thus perform reordering the list before loading the prefabs. This would probably help solve exclusion list issues, such as missing managers. But the purpose of automatic dependency would be to release that constraint from the users (as they are probably not aware of these dependencies in the first place)
> 
> for example: `[ManagerDependsOn(typeof(UIEventManager),typeof(GameManager))]`
> 
Yes, this is a good idea. I had a problem implementing a UI Tooltip Manager that needs on Awake event to access the InputSystemUIModule component loaded by the UIEventManager:

            if (TryGet<UIEventManager>(out var uiEventManager))
            {
                _inputModule = uiEventManager.GetComponent<InputSystemUIInputModule>();
            }
            else
            {
                Debug.LogError("You need to start UIEventManager before UITooltipManager");
            }

> Oh, so it seems that it's an editor-only feature? If I understand correctly your use case, you want to do local changes before applying them to the main prefab? In that case I have some suggestions/concerns:
> 
> * When we are in a "scene overridden" mode, it can be difficult to tell which managers are not using the default prefabs, thus behaviour could be different between in-editor play mode and built players. I am a bit concerned about that as it would probably require a clear feedback to view these changes, and also be able to test it in both conditions.
> * Performing editor-only changes / testing requires if possible as few changes to the main runtime code. While it's not always possible (for example, I had to add a `EditorLogic` to handle correctly Play from Here). The rule of thumb is to leave the runtime code unaffected, for example by adding extra editor code between `#if UNITY_EDITOR` directives.
> 

A screenshot is better than 1000 words :) 

![Immagine 2021-12-23 181541](https://user-images.githubusercontent.com/666841/147273081-b8010ceb-a06f-43f7-94c3-1aa26d612c42.png)

This is my URP hierarchy project to test features. In the _DEMOS/Models directory there is a SCN_Models that by default loads everything from Assets/Resources. 
In the _DEMOS/StarshipController there is a directory Resources with a custom prefab DayNightManager that overrides the main Assets/Resources/DayNightManager and a custom Ingredients Settings (because in this scene I want to enable the debug).
All the other prefabs (UIEventManager, GameManager etc, are loaded from Assets/Resources) 
I did this because in that scene I want to setup my Day Night Manager with different parameters and calculate the position of the sun from a different latitude/longitude/date/time. This feature help me to avoid each time to make a copy of prefabs or work in a new project. 

> 
> Please note that your additions are pretty interesting, I am absolutely not against these but these require a bit more work in order to integrate safely. I don't think it's much work and I can also help you on that if you want. However if you feel that it's something that you can't handle, I can prioritize changes on my side and try to start branches to help you.

Cool I will work on your suggestions and give you a feedback soon. Merry Christmas Thomas and see you soon!